In [16]:
import numpy as np
import pandas as pd
import random

import matplotlib.pyplot as plt
import seaborn as sns

from kaggle_environments import make, evaluate

In [18]:
%%writefile rock_agent.py
# Agent that always plays rock
def rock_agent(observation, configuration):
    return 0  # 0 - rock

Overwriting rock_agent.py


In [20]:
%%writefile copy_opponent.py
import random
# Agent that copies the opponent's last action
def copy_opponent(observation, configuration):
    if observation.step > 0:
        return observation.lastOpponentAction
    else:
        return random.randrange(0, configuration.signs)

Overwriting copy_opponent.py


In [22]:
%%writefile random_agent.py
import random
# Agent that plays randomly
def random_agent(observation, configuration):
    return random.randrange(0, configuration.signs)  # Randomly choose 0, 1, or 2

Overwriting random_agent.py


In [24]:
%%writefile paper_agent.py
# Agent that always plays paper
def paper_agent(observation, configuration):
    return 1  # 1 - paper

Overwriting paper_agent.py


In [26]:
%%writefile scissors_agent.py
# Agent that always plays scissors
def scissors_agent(observation, configuration):
    return 2  # 2 - scissors


Overwriting scissors_agent.py


In [36]:
from kaggle_environments import make
import itertools
# Создаем окружение
env = make("rps")
# Список всех агентов
agents = [
    "rock_agent.py",
    "copy_opponent.py",
    "random_agent.py",
    "paper_agent.py",
    "scissors_agent.py"
]
# Словарь для хранения результатов
results_summary = {agent: {"wins": 0, "losses": 0, "draws": 0} for agent in agents}
# Количество эпизодов для каждой пары
num_episodes = 100

# Запускаем все возможные пары агентов
for agent1, agent2 in itertools.combinations(agents, 2):
    results = env.run([agent1, agent2])
    print(f"Results for {agent1} vs {agent2}:")
    for i, result in enumerate(results):
        print(f"  Episode {i + 1}: {result}")

Results for rock_agent.py vs copy_opponent.py:
  Episode 1: [{'action': 0, 'reward': 0, 'info': {}, 'observation': {'remainingOverageTime': 60, 'step': 0, 'reward': 0}, 'status': 'ACTIVE'}, {'action': 0, 'reward': 0, 'info': {}, 'observation': {'remainingOverageTime': 60, 'reward': 0}, 'status': 'ACTIVE'}]
  Episode 2: [{'action': 0, 'reward': -1.0, 'info': {}, 'observation': {'remainingOverageTime': 60, 'step': 1, 'reward': -1, 'lastOpponentAction': 1}, 'status': 'ACTIVE'}, {'action': 1, 'reward': 1.0, 'info': {}, 'observation': {'remainingOverageTime': 60, 'reward': 1, 'lastOpponentAction': 0}, 'status': 'ACTIVE'}]
  Episode 3: [{'action': 0, 'reward': -1.0, 'info': {}, 'observation': {'remainingOverageTime': 60, 'step': 2, 'reward': -1, 'lastOpponentAction': 0}, 'status': 'ACTIVE'}, {'action': 0, 'reward': 1.0, 'info': {}, 'observation': {'remainingOverageTime': 60, 'reward': 1, 'lastOpponentAction': 0}, 'status': 'ACTIVE'}]
  Episode 4: [{'action': 0, 'reward': -1.0, 'info': {}, 'o

После того, как мы запустили всевозможные пары агентов, посчитаем количество побед для каждой пары и посмотрим какая стратегия играет лучше в нашем случае. 

In [37]:
from kaggle_environments import make
import itertools
# Создаем окружение
env = make("rps")
# Список всех агентов
agents = [
    "rock_agent.py",
    "copy_opponent.py",
    "random_agent.py",
    "paper_agent.py",
    "scissors_agent.py"
]
# Словарь для хранения результатов
results_summary = {agent: {"wins": 0, "losses": 0, "draws": 0} for agent in agents}
# Количество эпизодов для каждой пары
num_episodes = 100
# Запускаем все возможные пары агентов
for agent1, agent2 in itertools.combinations(agents, 2):
    print(f"Запуск: {agent1} против {agent2}")
    
    # Запускаем многократные эпизоды для каждой пары агентов
    for _ in range(num_episodes):
        # Получаем результаты игры
        results = env.run([agent1, agent2])
        # Результаты содержат информацию о действиях агентов
        agent1_action = results[-1][0]['action']  # Действие первого агента
        agent2_action = results[-1][1]['action']  # Действие второго агента
        # Определяем победителя
        if agent1_action == agent2_action:
            results_summary[agent1]["draws"] += 1
            results_summary[agent2]["draws"] += 1
        elif (agent1_action == 0 and agent2_action == 2) or \
             (agent1_action == 1 and agent2_action == 0) or \
             (agent1_action == 2 and agent2_action == 1):
            results_summary[agent1]["wins"] += 1
            results_summary[agent2]["losses"] += 1
        else:
            results_summary[agent1]["losses"] += 1
            results_summary[agent2]["wins"] += 1
# Выводим результаты
print("\nРезультаты:")
for agent, result in results_summary.items():
    print(f"{agent}: Победы: {result['wins']}, Поражения: {result['losses']}, Ничьи: {result['draws']}")

Запуск: rock_agent.py против copy_opponent.py
Запуск: rock_agent.py против random_agent.py
Запуск: rock_agent.py против paper_agent.py
Запуск: rock_agent.py против scissors_agent.py
Запуск: copy_opponent.py против random_agent.py
Запуск: copy_opponent.py против paper_agent.py
Запуск: copy_opponent.py против scissors_agent.py
Запуск: random_agent.py против paper_agent.py
Запуск: random_agent.py против scissors_agent.py
Запуск: paper_agent.py против scissors_agent.py

Результаты:
rock_agent.py: Победы: 124, Поражения: 139, Ничьи: 137
copy_opponent.py: Победы: 24, Поражения: 33, Ничьи: 343
random_agent.py: Победы: 137, Поражения: 124, Ничьи: 139
paper_agent.py: Победы: 140, Поражения: 129, Ничьи: 131
scissors_agent.py: Победы: 136, Поражения: 136, Ничьи: 128


Вывод:

Наилучшие результаты показывает агент paper_agent.py, который одержал 140 побед и имеет минимальное количество поражений (129). Это делает его наиболее успешным среди протестированных агентов.

Агент copy_opponent.py демонстрирует наименьшую эффективность, с всего лишь 24 победами, несмотря на большое количество ничьих (343). Это может указывать на то, что он часто повторяет ходы противников, что в данной стратегии не всегда приводит к победам.

Агенты scissors_agent.py и random_agent.py также показывают хорошие результаты, с 136 и 137 победами соответственно, но не достигают уровня paper_agent.py.

Агент rock_agent.py имеет 124 победы и 139 поражений, что делает его менее успешным по сравнению с другими агентами в данной конфигурации.

Таким образом, стратегия, использующая "бумагу" (paper), оказывается самой эффективной в данной конфигурации игры "камень, ножницы, бумага".